Improve torch.cond useability: Return UserError with actionable error messages

guangy10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98909

It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

guangy10 committed:
Improve torch.cond useability: Return UserError with actionable error messages

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
Instead of having a faux error code in the UserError message, maybe you should define a hierarchy of errors below UserError?

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -923,14 +966,18 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+                except TypeError as e:
+                    raise UserError(
+                        UserErrorType.COND_OP_RESTRICTION,
+                        f"{UserErrorType.COND_OP_RESTRICTION.name}. {str(e)}",
```
oh wait, you store the code in UserError. So why doesn't UserError print it directly as part of rendering 🤔

guangy10 replied:
Yep. Will address this

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, in general `true_fn` and `false_fn` (defined by users) must return a single tensor with exact same metadata, e.g. shape, dtype, etc.
if one branch returns a single tensor but the other branch return 2 tensors, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect true branch and false branch return with same length but got
  true_fn returns [sin]
  false_fn returns [x_1, x_1]
```
If both branches are returning a single tensor but with different shapes, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect each tensor returned from true branch and false branch has exact same metadata but got
  true_fn returns TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  false_fn returns TensorMetadata(shape=torch.Size([10, 10]), dtype=torch.float32, requires_grad=False, stride=(10, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

Another example about `operands`. In general, it must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect operands to be a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented on functorch/experimental/_cond.py:
```diff
@@ -56,12 +56,35 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
     flat_false_outs, _ = pytree.tree_flatten(false_outs)
-    assert(len(flat_true_outs) == len(flat_false_outs))
+    if len(flat_true_outs) != len(flat_false_outs):
+        raise UserError(
```
I am not sure if it is a good idea to use UserError here as it is kind of TorchDynamo specific thing. Should we consider moving UserError to more higher level place? cc @yanboliang @anijain2305 @avikchaudhuri

tugsbayasgalan replied:
Maybe you can just throw different type of Error subclass here and use that to raise UserError in dynamo side?

avikchaudhuri replied:
I do think this is a user error: outputs need to have the same types in both branches, which is a standard thing to require in other settings as well. 

I do worry about UserError / UserErrorType becoming a hot mess over time, though. E.g., COND_OP_RESTRICTION seems too low-level as a classifier. Something like CONTROL_FLOW_RESTRICTION would be better to cover more.

guangy10 replied:
> I do think this is a user error: outputs need to have the same types in both branches, which is a standard thing to require in other settings as well.

The question is do we expect users to take actions when seeing this error. Isn't still user who write these branches? Currently I revert it back to assert as you guys think it's too obvious to point out that outputs need to have the same length and meta in both branches.

avikchaudhuri replied:
To be clear, I said I think these *should* be user errors, because I think the user is in control of metadata of the outputs; OTOH I guess @tugsbayasgalan is saying that exposing internal data structures to users to assert their equality across branches is not good, which I also agree with, so we just have to check equality of each part separately to describe the problem in terms of what the user can understand.

guangy10 replied:
Oh, I read it as it should not! My bad.

guangy10 replied:
Addressed

avikchaudhuri commented on functorch/experimental/_cond.py:
```diff
@@ -56,12 +56,35 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+    if len(flat_true_outs) != len(flat_false_outs):
+        raise UserError(
+            UserErrorType.COND_OP_RESTRICTION,
+            "Expect true branch and false branch return with same length but got\n"
```
This message is not very clear.
1. Grammatic fix: "Expected ... but got ..."
2. "return with same length" -> "to return same number of outputs"
3. No point emitting `true_outs` and `false_outs` (could be large!), just emit their `len()`?

avikchaudhuri commented on functorch/experimental/_cond.py:
```diff
@@ -56,12 +56,35 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+        if true_out.meta['tensor_meta'] != false_out.meta['tensor_meta']:
+            raise UserError(
+                UserErrorType.COND_OP_RESTRICTION,
+                "Expect each tensor returned from true branch and false branch "
```
Expect[ed] each tensor...[to have] [CUT exact] same metadata but [CUT got]

tugsbayasgalan replied:
Is metadata user exposed concept? What would user even do from seeing metadata? Maybe this should just be an assertion.

avikchaudhuri replied:
Hmm, you're right that metadata is not really something the user needs to understand. But each of the individual components of metadata are, right? Like, shape, dtype, etc. are things user should understand, and it's OK to complain when they don't match across branches.

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5),)
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5),)
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5),)
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5),)
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5), torch.rand(2))
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.rand(5),)
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2455,6 +2455,104 @@ def f_pred_traced_as_tensor_var(x):
+        example_inputs = (torch.randn(4), torch.randn(2))
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -389,11 +389,15 @@ def f(x, y):
-        with self.assertRaises(AssertionError):
+        with self.assertRaisesRegex(
+            UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -404,7 +408,11 @@ def f(x, y):
-        with self.assertRaises(AssertionError):
+        with self.assertRaisesRegex(
+            UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -551,11 +559,15 @@ def f(x, y):
-        with self.assertRaises(AssertionError):
+        with self.assertRaisesRegex(
+            UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on test/functorch/test_control_flow.py:
```diff
@@ -566,7 +578,11 @@ def f(x, y):
-        with self.assertRaises(AssertionError):
+        with self.assertRaisesRegex(
+            UserError,
+            f"{torch._dynamo.exc.UserErrorType.COND_OP_RESTRICTION.name}: "
```
cut

avikchaudhuri commented on torch/_dynamo/exc.py:
```diff
@@ -96,6 +97,10 @@ def __init__(self, error_type: UserErrorType, msg):
+        self.message = msg
+
+    def __str__(self):
+        return f"{self.error_type.name}: {self.message}"
```
I'm not sure this is a good idea. The error type should be an internal classifier: it doesn't matter to the user what kind of error it is. At most, I'd be OK with the user seeing <SomeType>UserError, but never UserError(<SomeType). I'd also be OK with some numeric error code that doesn't detract from the error message and that we don't have to keep tweaking the name of as we refactor / rename error types.

avikchaudhuri commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2035,15 +2035,14 @@ def inline_call_(
-                kwargs,
-                e,
+            raise Unsupported(
+                "{reason}.\n  func = {func}, args = {args}, kwargs = {kwargs}".format(
```
Mmm, I don't think returning `func` `args` `kwargs` is that helpful. Maybe look up frame summaries and current loc from the tracing context instead.

guangy10 replied:
The error here is because of signature mismatch. I'm not sure why it's not helpful by showing where the function is defined (code location with line number) and what `args` are passed to it. 

It's wrapped to a UserError, so the message is like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = <code object false_fn at 0x7fe7bfd36fa0, file "/scratch/guangyang/work/pytorch/test/dynamo/test_export.py", line 2514>, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```

avikchaudhuri replied:
Sorry, what I meant is that returning the raw objects is not helpful. What I was suggesting you do instead is that you grab pre-formatted stack summaries from the tracing context, which should give you the function name / line number as well, but as a standard Python traceback instead of with funny `<code object...>`.

guangy10 replied:
I see. That makes sense to me

guangy10 replied:
Done

avikchaudhuri commented on torch/_dynamo/variables/torch.py:
```diff
@@ -855,12 +855,20 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
-            assert isinstance(output, TensorVariable)
+            if not isinstance(output, TensorVariable):
+                raise TypeError(
+                    "Expect branch out type to be a single tensor but got {}".format(
```
Expect[ed]

avikchaudhuri commented on torch/_dynamo/variables/torch.py:
```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            if type(args[0]) not in (TensorVariable, SymNodeVariable, ConstantVariable):
+                raise UserError(
+                    UserErrorType.COND_OP_RESTRICTION,
+                    f"Expect pred to be traced as TensorVariable, "
```
Expect[ed]

avikchaudhuri commented on torch/_dynamo/variables/torch.py:
```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            if type(args[3]) is not ListVariable:
+                raise UserError(
+                    UserErrorType.COND_OP_RESTRICTION,
+                    f"Expect operands to be a list but got {args[3].python_type()}",
```
Expect[ed] a list [CUT operands] but got...

avikchaudhuri commented on torch/_dynamo/variables/torch.py:
```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            ):
+                raise UserError(
+                    UserErrorType.COND_OP_RESTRICTION,
+                    "Expect operands to be a list of tensors but got {actual_args}".format(
```
Expect[ed] a list of tensors but got...

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, in general `true_fn` and `false_fn` (defined by users) must return a single tensor with exact same metadata, e.g. shape, dtype, etc.
if one branch returns a single tensor but the other branch return 2 tensors, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect true branch and false branch return with same length but got
  true_fn returns [sin]
  false_fn returns [x_1, x_1]
```
If both branches are returning a single tensor but with different shapes, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect each tensor returned from true branch and false branch has exact same metadata but got
  true_fn returns TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  false_fn returns TensorMetadata(shape=torch.Size([10, 10]), dtype=torch.float32, requires_grad=False, stride=(10, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

Another example about `operands`. In general, it must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: COND_OP_RESTRICTION. Expect operands to be a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = <code object 'false_fn' at 0x7fe7bfd36fa0, file "~/pytorch/test/dynamo/test_export.py", line 2514>, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.Unsupported: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -837,12 +837,15 @@ def graph_with_interpreter(*args):
+                    tracing_mode="real",
+                    _allow_non_fake_inputs=True,
+                )(*example_fake_inputs)
+            except TypeError as e:
```
This looks wrong. You have no reason to believe the TypeError is specifically because of dynamic control flow. Classify the errors at raise site, not here.

guangy10 replied:
Addressed by defining and catching cond specific exceptions

ezyang commented on torch/_dynamo/variables/functions.py:
```diff
@@ -95,9 +95,12 @@ def get_name(self):
+                self, list(self.self_args()) + list(args), kwargs
+            )
+        except Exception:
+            raise
```
?

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
-            )  # predicate
+            if len(args) != 4:
+                raise UserError(
+                    UserErrorType.DYNAMIC_CONTROL_FLOW,
```
I mentioned this in someone else's PR, but I'm not sure why you don't have a error hierarchy, instead of splitting it into an enum + error name.

guangy10 replied:
I think we want only want to the very high level `UserError` to the users with a concise and actionable message. No need to expose detailed user error types, which is helpful for us but won't be useful for user. @avikchaudhuri mentioned similar point in a different comment.

guangy10 replied:
BTW, we just want a flat user-facing error, but we can still define internal error hierarchy to throw and catch specific errors and wrap those to the user-facing errors when needed.

ezyang replied:
Err, concretely, I'm proposing DynamicControlFlowUserError(...) analog here. This doesn't seem overly detailed; it seems descriptive!

guangy10 replied:
From users perspective, having `DynamicControlFlowUserError` or `UserError.DynamicControlFlowError` doesn't make much different. I'm okay with ether way.

However, from developers perspective, I think we should keep either the user error hierarchy or user error type enum, but never do both. So I would propose keep using error type enum in this PR as it's already in use in multiple places in the code. If we agree to define a user error hierarchy going forward, we can have a follow-up PR migrating all use cases to it. 

Let me know your thoughts. CC: @ezyang @tugsbayasgalan @avikchaudhuri

avikchaudhuri replied:
From the user's perspective, the most significant bit is whether something is a user error or an internal error. So that distinction is all the matters (along with the promise that all user errors have a high bar for actionable message, suggesting a fix / linking to an example).

I do think some kind of hierarchy is needed, otherwise we'll just end up with a "dumping ground" for all kinds of user error types without any organizing principle. Moreover, I like the idea of using subclasses because then we can add more fields specific to some kinds of user errors that are not applicable to others.

Now, what's the point of having an enum? To me, it's about having a flat list of error codes that we can use to collect analytics: how often do users hit some error code, how often do they search for a workaround given an error code, etc. The hierarchy is purely for managing our codebase.

So to have the best of both worlds, I'd suggest we move to a class-based hierarchy where the base class has a static error code (implemented as a monotonically increasing enum) along with any other common required fields. We print the error code as a prefix of the error message (something like `user error [code]`) to signal that it is a user error, but can otherwise take any other fields as well.

guangy10 replied:
Agree with most part. I think having user error hierarchy is sufficient to support both:
 1. better codebase management for error raising
 2. gather analytics. As long as the error can be aggregated, we will see things like 20% of users errors due to `DynamicControlFlowArgMismatch`, etc.

The error type enum (error code) can be optionally defined for each error in the hierarchy.

I may think differently about exposing error code to users, which is neither user understandable nor actionable w/o showing the meaning of it. The fact is, the meaning of it is already covered in the actionable error message.

Refactoring UserError to a user error hierarchy (like internal errors) will be addressed in a separate PR.

guangy10 replied:
Will start merging we agree on the actions.

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -855,12 +855,20 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            try:
+                output = f.call_function(tx, args, {})
+            except Exception:
+                raise
```
This try-except seems pretty pointless?

guangy10 replied:
I understand that python can implicitly propagate exceptions up to the higher layer, but I do also see dynamo code where we make the propagation explicitly like in c++ for readability I guess. Will revert if not preferred.

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -855,12 +855,20 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+            if not isinstance(output, TensorVariable):
+                raise TypeError(
+                    "Expected branch out type to be a single tensor but got {}".format(
+                        str(output.python_type())
```
nit: you don't need to str here, it will be implicitly str'ed by the format operation. Also, use an f-string, much nicer:
`f"Expected branch out type to be a single tensor but got {output.python_type()}"`

ezyang replied:
It's not clear to me why this is a TypeError and not a UserError, under the logic for UserError? As in, not returning a Tensor from the subgraph is clearly a problem for the user to solve?

Also, if we're modifying this code, you may want to distinguish the case for not supporting multi-Tensor output (tuple/list return), since that is not a UserError.

guangy10 replied:
Yeah, it is a UserError. Keep it as an internal error here just because this function is not only used by cond(), so this exception is propagated up and is wrapped to be a UserError at the entry of cond().

ezyang replied:
This seems fairly error prone in the long term? How are you going to get people to remember to use UserError or an internal error whenever they are writing errors? How are you going to make sure people consistently do the wrapping? (And how are you going to deal with people who are complaining that rewrapping exceptions blows up the stack trace?)

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -888,14 +896,45 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+                raise UserError(
+                    UserErrorType.DYNAMIC_CONTROL_FLOW,
+                    f"Expected pred to be traced as TensorVariable, "
+                    f"SymNodeVariable or ConstantVariable but got {str(type(args[0]))} "
```
TensorVariable, SymNodeVariable and ConstantVariable are meaningless to end users. They aren't going to know anything about how Dynamo internally structures its VariableTracker hierarchy.

guangy10 replied:
Yeah, rephrase the error message in a more user understandable way

tugsbayasgalan commented on functorch/experimental/_cond.py:
```diff
@@ -56,12 +56,22 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+    if len(flat_true_outs) != len(flat_false_outs):
+        raise TypeError(
+            f"Expected to return same number of outputs but got:"
+            f"\n  {true_fn.__name__} returns {flat_true_outs}"
```
Could you just return the lengths of the outputs? I feel like that would be more informative to the user?

guangy10 replied:
Sure.

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.Unsupported: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@guangy10 FYI I'm tracking another known limitation here: https://github.com/pytorch/pytorch/pull/99367

guangy10 commented:
OK. doctr_det_predictor failures in the CI are irrelevant.

guangy10 commented:
@pytorchbot merge -ic

guangy10 commented:
The macos issue on trunk doesn't seem to be relevant.

guangy10 commented:
@pytorchbot merge -ic

guangy10 commented:
@pytorchbot merge -ic -r master

guangy10 commented:
@pytorchbot merge -ic -r main

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented:
Resolve merge conflicts. Linter errors

guangy10 commented:
@pytorchbot merge

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented:
Reverted #98900 which doesn't make the common pred expression (e.g. `pred = x.dim() > 2 and x.shape[1] > 5`) being traced correctly. Updated test cases to show the workaround, that is users will have to either:
 1. Change the order of the expression `pred = x.shape[1] > 5 and x.dim() > 2` (This isn't intuitive and doesn't make much sense logically)
 2. Make the `pred` type be `torch.SymBool` explicitly. (This is what the updated test cases and exportdb example show)

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented on test/dynamo/test_export.py:
```diff
@@ -2448,22 +2448,29 @@ def true_fn(x):
-        example_inputs = (torch.rand(5),)
+        def f_pred_complex_expression_traced_as_symnode_var(x):
+            return cond(
+                x.dim() > 1 and x.shape[1] > 5 and x.shape[1] <= 10,
```
@tugsbayasgalan It turns out this is just traced fine:
```
test_cond_supported_pred_types_dynamic_shapes (torch._dynamo.testing.DynamicShapesExportTests) ...


def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    sym_size = torch.ops.aten.sym_size(arg0, 1)
    le = sym_size <= 10;  sym_size = None
    true_graph_0 = self.true_graph_0
    false_graph_0 = self.false_graph_0
    cond = torch.ops.cond(le, true_graph_0, false_graph_0, [arg0]);  le = true_graph_0 = false_graph_0 = arg0 = None
    return pytree.tree_unflatten([cond], self._out_spec)
```

guangy10 replied:
@tugsbayasgalan Maybe not completely correct. It looks like the condition is over simplified. But anyway the issue of predicate became specialized doesn't seem to be relevant to #98900. And I do see more similar cases in the exportdb when the predicate is simply a shape check (https://www.internalfb.com/intern/staticdocs/exportdb/cond_branch_nested_function.html). CC: @angelayi

guangy10 committed:
Update on "Improve torch.cond useability: Return UserError with actionable error messages"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by separating user errors from the dynamo internal errors. By definition, user error means the usage of `torch.cond()` violates the restrictions of this API therefore needs users to take action and fix the error.

In this notebook N3363227 we discovered a bunch of limitations of using `torch.cond(pred, true_fn, false_fn, operands)`. In summary, the limitations can be categorized as:
 - predicate restriction (`pred`)
 - operands restriction (`operands`)
 - branch restriction (`true_fn` & `false_fn`)

The error message will be more accurate about where the (user) error is from and more actionable for users to fix it.

For example, `operands` must be a list of tensors and the signature of `true_fn` and `false_fn` must match with the `operands`.
If the operands contains non-tensor types, user will see error message like:
```
torch._dynamo.exc.UserError: Expected a list of tensors but got ["<class 'torch.Tensor'>", "<class 'float'>"]

from user code:
   File "~/pytorch/test/dynamo/test_export.py", line 2504, in f_non_tensor_operands
    return cond(True, lambda x, a: x.sin(), lambda x, a: x.cos(), [x, a])
```
If the signature of the branch function doesn't match with `operands`, user will see error message like:
```
torch._dynamo.exc.UserError: too many positional arguments.
  func = 'false_fn' ~/pytorch/test/dynamo/test_export.py:2514, args = [<class 'torch.Tensor'>, <class 'torch.Tensor'>], kwargs = {}
```
Or if the tensor returned from user defined branches has different metadata, e.g. shapes, dtypes, etc., user will see error message like:
```
TypeError: Expected each tensor to have same metadata but got:
  cond_true_0 returns TensorMetadata(shape=torch.Size([2, 1]), dtype=torch.int64, requires_grad=False, stride=(1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
  cond_false_0 returns TensorMetadata(shape=torch.Size([1]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={})
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented:
@pytorchbot merge
